ui/src/dataExplorer/components/SaveAsTaskForm.tsx

```diff
@@ -109,14 +103,58 @@ class SaveAsTaskForm extends PureComponent<Props & WithRouterProps> {
           onSubmit={this.handleSubmit}
           canSubmit={this.isFormValid}
           dismiss={dismiss}
-          tokens={tokens}
+          tokens={this.getRelevantTokens()}
           selectedToken={selectedToken}
           onTokenChange={this.handleTokenChange}
         />
       </SpinnerContainer>
     )
   }
 
+  private getRelevantTokens() {
+    const authorizations = this.props.tokens
+
+    const {draftQueries, activeQueryIndex} = this.props
+    let readBucketName = ''
+    if (draftQueries[activeQueryIndex].editMode === 'builder') {
+      readBucketName = draftQueries[activeQueryIndex].builderConfig.buckets[0]
+    } else {
+      const text = draftQueries[activeQueryIndex].text
+      const splitBucket = text.split('bucket:')
+      const splitQuotes = splitBucket[1].split('"')
+      readBucketName = splitQuotes[1]
+    }
+    const writeBucketName = this.props.taskOptions.toBucketName
+
+    const readAuthorizations = authorizations.filter(auth =>
+      auth.permissions.some(
+        permission =>
+          permission.action === 'read' &&
+          permission.resource.type === 'buckets' &&
+          (!permission.resource.name ||
+            permission.resource.name === readBucketName)
+      )
+    )
+
+    const writeAuthorizations = authorizations.filter(auth =>
+      auth.permissions.some(
+        permission =>
+          permission.action === 'write' &&
+          permission.resource.type === 'buckets' &&
+          (!permission.resource.name ||
+            permission.resource.name === writeBucketName)
+      )
+    )
+
+    const relevantAuthorizations = _.intersectionBy(
+      readAuthorizations,
+      writeAuthorizations,
+      'id'
+    )
+
+    return relevantAuthorizations
```

Comment:
There's a lot going on here.  Perhaps it would be best to extract each of these steps here into separate helper functions and unit test those functions?